_drafts/Firmware-encryption-with-python.md

```diff
@@ -0,0 +1,643 @@
+
+```c
+void image_decrypt(void) {
+    uint32_t slot_Cipher_appAddr = SLOT_ADDRESS_CIPHER_APP;
+    uint32_t slot_Plain_appAddr = SLOT_ADDRESS_PLAIN_APP;
+    int8_t idx = 3;
+    uint32_t word;
+    int32_t bytes_remaining = FIRMWARE_PLAIN_APP_SIZE;
+    
+    word = *(uint32_t*)(slot_Cipher_appAddr);
+    
+    flash_unlock();
+    //! Erase flash sector slot app
+    flash_erase(2, 2);
+    while(bytes_remaining) {
+        //! Read 16 bytes of encrypted data
+        for(size_t i=0; i<NUMBER_BYTES_TO_PROCESS; i++) {
+            if(idx < 0) {
+                idx = 3;
+                slot_Cipher_appAddr += 4;
+                word = *(uint32_t*)(slot_Cipher_appAddr);
+            }
+            CipherFirmware[i] = (word >> (8*idx--)) & 0xff;
+        }
+        //! Decrypt the 16 bytes of data firmware
+        aes_decryption( CipherFirmware, NUMBER_BYTES_TO_PROCESS, PlainFirmware );
+        //! Write 16 bytes of data to flash memory
+        image_install( PlainFirmware, &slot_Plain_appAddr);
+        bytes_remaining -= 16;
+        idx = 3;
+        slot_Cipher_appAddr += 4;
+        word = *(uint32_t*)(slot_Cipher_appAddr);
+        if(bytes_remaining <= 0 ) bytes_remaining = 0;
+    }
+    flash_lock();
+}
+```
+
+Finally, upon completing the decryption and installation of the firmware image, the boot will launch the application on every restart. This ensures that the decrypted firmware is executed and the device operates using the updated firmware:
+
+```c
+void image_start(void) {
+    uint32_t jump_address;
+    typedef void(*start_app)(void);
+    start_app p_jump_app;
+    jump_address = *(volatile uint32_t *)(((uint32_t)SLOT_ADDRESS_PLAIN_APP + 4));
+    p_jump_app   = (start_app)jump_address;
+    //! Initialize loader's Stack Pointer 
+    __set_MSP(*(__IO uint32_t *)(SLOT_ADDRESS_PLAIN_APP));
+    //! Jump into app
+    p_jump_app();
+}
+```
+## Closing
+
+In conclusion, while encrypting the firmware image is a step towards enhancing security, it is not sufficient on its own. When deploying OTA updates, it is crucial to consider a combination of authenticity and confidentiality measures. Implementing firmware encryption requires additional effort, but it is more effective when paired with firmware image signing and robust flash memory protection on the MCU.
+
+Furthermore, proper key management is essential. Storing encryption keys in a secure element ensures their integrity and prevents unauthorized access. By adopting a comprehensive approach to firmware security, including encryption, image signing, flash memory protection, and secure key management, we can significantly enhance the overall security of our firmware updates.
```

Comment:
I was going to suggest touching on secure element a bit more for key storage. You don't need to write an example for it, but it might be nice to touch on it more in the key storage section. Right now it seems like we're storing the key in the firmware, which is fine for the example, but we should touch on why that wouldn't be right for production.

Reply:
Thanks for the feedback. Make the changes mentioned, I look forward to your comments again.